### PR TITLE
Make chat ignore permanent

### PR DESCRIPTION
**Pull request by Dune-jr**
Reimplements https://github.com/teeworlds/teeworlds/pull/2043 (but without the useless virtual inheritance OOP mindfuck):

> **Note**: this slightly conflicts with #2042 (since it modifies the mute feature)
> 
> ### Behaviour
> This feature makes use of the existing friends system to implement a permanent chat ignore. It adds two console commands `add_ignore` and `remove_ignore` (feel free to suggest different wording...)
> 
> * Muting or unmuting a player, from the console or the menus, is always permanent (and player+clan based)
> * An information message is displayed when
>   
>   * a previously muted player join the server
>   * the player joins a server with previously muted players
> 
> `%s is muted by you` sounds like broken English, maybe @oy will have a better suggestion :)
> image
> 
> ### Implementation
> As mentioned above, this feature makes use of the `IFriends` and `CFriends` classes.
> 
> In order to register two IFriends interfaces, I ran into the ["diamond problem"](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) where a class inherits of two classes which themselves inherit from the same baseclass. This is solved using virtual inheritence:
> 
> ```c++
> class IFriends: public IInterface
> class IGoodFriends: public virtual IFriends
> class IBadFriends: public virtual IFriends
> class CFriends : public virtual IFriends
> class CGoodFriends: public CFriends, public IGoodFriends
> class CBadFriends: public CFriends, public IBadFriends
> ```
> 
> Multiple inheritance is generally frowned upon, but it seems like it tends to be given a pass for interfaces.
> 
> Due to this, there can be some nasty runtime errors if you use a `static_cast` instead of a `reinterpret_cast` (you get a cryptic "virtual baseclass botch" error) in gameclient.h.
> 
> It seemed somewhat clever and avoids code redundancy but I'm as much of an OOP nerd as my neighbor's cat is a leopard and I won't mind if I have to scrap that and find another way, given that it is somewhat obscure and possibly unfitting for the Teeworlds basecode.

The class diagram now looks like this:

image

It still works fine, even after quitting the game:
screenshot_2019-09-14_20-25-42

I had to rename a lot of `Friends` stuff to `Contacts`, so the diff file is a bit unreadable. Probably should have done this rename in a separate commit :(
I am hesitating to rename the three `friends.h` files to `contacts.h`, might be more fitting as well.

### CODE REVIEW

**Comment by fokkonaut** [comment]
Can we have a command like friends_ignore_clan for this? So that friends and muted players is only based on name? We had this in DDNet, and it was pretty good, because people change clans frequently, for fun or if they are in multiple clans.

**Comment by Dune-jr** [comment]
@fokkonaut I agree with this sentiment, but believe it should be done out of score of this PR, and also affect friendship.

For example, if I add as a friend "AMN", I believe it should mark "[5%] AMN" as my friend. However, if my add "[5%] AMN", it should probably not mark "AMN" as a friend. That is my take on this.

File: src/game/client/components/menus.cpp [hunk 1]
```diff
@@ -9,12 +9,12 @@
 #include <engine/config.h>
 #include <engine/editor.h>
 #include <engine/engine.h>
-#include <engine/friends.h>
 #include <engine/keys.h>
 #include <engine/serverbrowser.h>
 #include <engine/storage.h>
 #include <engine/textrender.h>
 #include <engine/shared/config.h>
+#include <engine/client/friends.h>
```
Comment (oy) [inline src/game/client/components/menus.cpp]: Hm, could you move the enums to the interface? It shouldn't include sth from `engine/client/` but work with the interfaces instead.

**Comment by oy** [comment]
Yeah, friends.* could be renamed to contacts.*

File: src/game/client/components/menus.cpp (hunk 1 above)
Comment (Dune-jr) [reply src/game/client/components/menus.cpp]: Yes, that makes sense

**Comment by Dune-jr** [comment]
Is it alright to leave the enums in the global namespace then?

**Comment by oy** [comment]
@Dune-jr better not put it in the global namespace.
Maybe add it to to CContactInfo?

**Comment by Dune-jr** [comment]
> @Dune-jr better not put it in the global namespace.
> Maybe add it to to CContactInfo?

I can, I don't see a better way anyway

> Edit:
> You have to update the cmake file list to fix the failed checks.

I actually did a `sed` and self-congratulated for thinking about cmake for once. But it broke the alphabetical order so nope :(

**Pull request merged**